src/common/utilities.cpp

```diff
@@ -0,0 +1,495 @@
+        const auto& dim = dimensions.at(i);
+        shard_strides.push_back(shard_strides.back() *
+                                zarr::common::shards_along_dimension(dim));
+    }
+
+    std::vector<size_t> shard_lattice_indices;
+    for (auto i = 0; i < dimensions.size(); ++i) {
+        shard_lattice_indices.push_back(chunk_lattice_indices.at(i) /
+                                        dimensions.at(i).shard_size_chunks);
+    }
+
+    size_t index = 0;
+    for (auto i = 0; i < dimensions.size(); ++i) {
+        index += shard_lattice_indices.at(i) * shard_strides.at(i);
+    }
+
+    return index;
+}
+
+size_t
+common::shard_internal_index(size_t chunk_idx,
+                             const std::vector<zarr::Dimension>& dimensions)
+{
+    // make chunk strides
+    std::vector<size_t> chunk_strides(1, 1);
+    for (auto i = 0; i < dimensions.size() - 1; ++i) {
+        const auto& dim = dimensions.at(i);
+        chunk_strides.push_back(chunk_strides.back() *
+                                zarr::common::chunks_along_dimension(dim));
+        CHECK(chunk_strides.back());
+    }
+
+    // get chunk indices
+    std::vector<size_t> chunk_lattice_indices;
+    for (auto i = 0; i < dimensions.size() - 1; ++i) {
+        chunk_lattice_indices.push_back(chunk_idx % chunk_strides.at(i + 1) /
+                                        chunk_strides.at(i));
+    }
+    chunk_lattice_indices.push_back(chunk_idx / chunk_strides.back());
+
+    // make shard lattice indices
+    std::vector<size_t> shard_lattice_indices;
+    for (auto i = 0; i < dimensions.size(); ++i) {
+        shard_lattice_indices.push_back(chunk_lattice_indices.at(i) /
+                                        dimensions.at(i).shard_size_chunks);
+    }
+
+    std::vector<size_t> chunk_internal_strides(1, 1);
+    for (auto i = 0; i < dimensions.size() - 1; ++i) {
+        const auto& dim = dimensions.at(i);
+        chunk_internal_strides.push_back(chunk_internal_strides.back() *
+                                         dim.shard_size_chunks);
+    }
+
+    size_t index = 0;
+
+    for (auto i = 0; i < dimensions.size(); ++i) {
+        index +=
+          (chunk_lattice_indices.at(i) % dimensions.at(i).shard_size_chunks) *
+          chunk_internal_strides.at(i);
+    }
+
+    return index;
+}
+
+size_t
+common::bytes_per_chunk(const std::vector<Dimension>& dimensions,
+                        const SampleType& type)
+{
+    auto n_bytes = bytes_of_type(type);
+    for (const auto& d : dimensions) {
+        n_bytes *= d.chunk_size_px;
+    }
+
+    return n_bytes;
+}
+
+const char*
+common::sample_type_to_dtype(SampleType t)
+
+{
+    static const char* table[] = { "u1", "u2", "i1", "i2",
+                                   "f4", "u2", "u2", "u2" };
+    if (t < countof(table)) {
+        return table[t];
+    } else {
+        throw std::runtime_error("Invalid sample type.");
+    }
+}
+
+const char*
+common::sample_type_to_string(SampleType t) noexcept
```

Comment:
Why does this function return a string if there’s an error (coupled with `noexcept`) and the previous one throws an exception?

Reply:
The intended use is different. `sample_type_to_dtype` is intended to write to Zarr metadata, so if the pixel type is unsupported, the program had better not silently write anything like "unrecognized pixel type". On the other hand, `sample_type_to_string` is simply used for logging, so the correct thing to do when pass an unrecognized `SampleType` is log that.